File: plugins/TaskManager/pwba_plugin.js
/*
+ ==============================================================================
|
|     TaskManager plugin JavaScript
|
+ ==============================================================================
*/

function ${P}() {
    let _self = this;

    _self.registered_tasks = [];
    _self.ready = false;

    _self.init = function()  // this is custom initialization for the specific plugin
    {
        pwba.log_msg('Custom initialization for "${P}" Plugin.');
        _self._get_registered_tasks();
    };

    _self._get_registered_tasks = function() {
        _self.plugin_to_python('get_registered_tasks', {});
    };

    _self.msg_levels = {
        INFO: 'info',
        DEBUG: 'debug',
        WARNING: 'warning',
        ERROR: 'error',
        CRITICAL: 'critical',
    };

    _self.test_fn = function()
    {
        alert('Hello from ${P}');
    };

    _self.to_html_safe_text = function (str) {
        return String(str).replace(/&/g, '&amp;')
                          .replace(/</g, '&lt;')
                          .replace(/>/g, '&gt;')
                          .replace(/"/g, '&quot;');
    };

    _self.msg_log_scroll_to_end = function(task_name)
    {
        let results = document.getElementsByClassName("${P}_" + task_name + "_log");
        if (results.length) {
            for (let r=0; r < results.length; r++) {
                let msg_log_div = results[r];
                let children = msg_log_div.children;

                var total_height = 80; // add in some height to adjust for padding and such
                for (let c=0; c < children.length; c++ ) {
                    total_height = total_height + children[c].offsetHeight;
                }
                msg_log_div.scrollTop = total_height;
            }
        }
    };

    _self.start_task = function(task_name, task_data) {
        if (! _self.ready) {
            console.log('"${P}" Plugin is not in a ready state yet. Unable to start task "' + task_name + '"');
            return;
        }

        if (_self.registered_tasks.indexOf(task_name) < 0) {
            // task_name not registered
            pwba.log_msg('[${P}] ERROR: Task "' + task_name + '" is not registered -- unable to start task.')
            return;
        }

        task_data.task_name = task_name;
        _self.plugin_to_python('start_task', task_data);
    };

    _self.set_progress_bar_percentage = function(task_name, percent_complete) {
        let el_results = document.getElementsByClassName("${P}_" + task_name + "_progress_fill");
        for (let c=0; c < el_results.length; c++) {
            let p_bar_div = el_results[c];
            p_bar_div.style.width = "" + percent_complete + "%";
        }
    };

    _self._set_controls_active_state = function(task_name, is_active) {
        let control_el_list = document.getElementsByClassName("${P}_" + task_name + "_control");
        for (let c=0; c < control_el_list.length; c++) {
            if (is_active) {
                if ('disabled' in control_el_list[c]) {
                    control_el_list[c].removeAttribute('disabled');
                }
            } else {
                control_el_list[c].setAttribute('disabled', 'disabled');
            }
        }
    };

    _self.disable_controls = function(task_name) { _self._set_controls_active_state(task_name, false); };

    _self.enable_controls = function(task_name) { _self._set_controls_active_state(task_name, true); };

    // -------------------------------------------------------------------------------
    //
    //  Below are functions called from Python, so each of these only take one
    //  argument "op_data" that is a dictionary/object with any data to pass along
    //
    // -------------------------------------------------------------------------------
    _self.received_registered_tasks = function(op_data) {
        _self.registered_tasks = op_data.registered_task_list;
        _self.ready = true;

        // DEBUG
        console.log('>>> in _self.received_registerd_tasks() function - registered tasks: ' +
                    _self.registered_tasks);
    };

    _self.ui_update_progress = function(op_data) {
        _self.set_progress_bar_percentage(op_data.task_name, op_data.percent_complete);
    };

    _self.ui_disable_controls = function(op_data) {
        _self.disable_controls(op_data.task_name);
    };

    _self.ui_message = function(op_data)
    {
        let task_name = op_data.task_name;

        let msg = op_data.message;
        let msg_level = op_data.message_level;

        let msg_level_class = "${P}_message_" + msg_level;

        if (msg_level == "info") {
            msg_line_prefix = "[INFO]: ";
        }
        else if (msg_level == "debug") {
            msg_line_prefix = "[DEBUG]: ";
        }
        else if (msg_level == "warn") {
            msg_line_prefix = "[WARNING]: ";
        }
        else if (msg_level == "error") {
            msg_line_prefix = "[ERROR]: ";
        }
        else if (msg_level == "critical") {
            msg_line_prefix = "[CRITICAL]: ";
        }
        else {
            msg_line_prefix = "[???]: ";
        }

        let results = document.getElementsByClassName("${P}_" + task_name + "_log");
        if (results.length == 0) {
            pwba.log_msg('[${P}] ERROR: No log DIV element found for Task "' + task_name + '" ... ' +
                         'unable to log output to UI.');
            return;
        }

        let msg_lines = msg.split('\n');

        for (let c=0; c < msg_lines.length; c++) {
            // output each line
            let m_line = msg_lines[c];
            if (c == 0)
                m_line = msg_line_prefix + m_line;

            let html_str = '<span class="' + msg_level_class +'">' + _self.to_html_safe_text(m_line) +
                            "</span><br/>";

            for (let r=0; r < results.length; r++) {
                let msg_log_div = results[r];
                msg_log_div.insertAdjacentHTML('beforeend', html_str);
            }
        }

        _self.msg_log_scroll_to_end(task_name);
    };

    _self.ui_task_ended = function(op_data)
    {
        let task_name = op_data.task_name;

        _self.set_progress_bar_percentage(task_name, 100.0);
        _self.msg_log_scroll_to_end(task_name);
        _self.enable_controls(task_name);

        setTimeout(function() {
            alert("Task '" + task_name + "' completed with status '" + op_data.completion_status +
                  '" and message "' + op_data.completion_message + '"')
        }, 100);
    };
}

pwba.register_plugin_js(${P});

File: plugins/TaskManager/pwba_plugin.py

import threading
import traceback

from PyWebBrowserApp import PluginBase
from PyWebBrowserApp import register_plugin_op

# + ============================================================================
# |
# |    TaskManager plugin Python class
# |
# + ============================================================================

class TaskStatuses(object):

    NOT_STARTED = 'not_started'
    WORKING = 'working'
    COMPLETED = 'completed'
    ERRORED = 'errored'
    PAUSED = 'paused'


class LogLevels(object):

    INFO = 'info'
    DEBUG = 'debug'
    WARNING = 'warning'
    ERROR = 'error'
    CRITICAL = 'critical'


class Task(object):

    def __init__(self, task_name, validation_to_start_fn, task_execution_fn, _ui_message_driver_fn,
                 _ui_update_progress_driver_fn, _ui_task_ended_driver_fn):

        self.task_name = task_name

        self.validation_to_start_fn = validation_to_start_fn
        self.task_execution_fn = task_execution_fn

        self._ui_message_driver_fn = _ui_message_driver_fn
        self.ui_update_progress_driver_fn = _ui_update_progress_driver_fn
        self.ui_task_ended_driver_fn  = _ui_task_ended_driver_fn

        self.execution_thread = None

        self.status = TaskStatuses.NOT_STARTED

    def get_status(self):

        return self.status

    def _task_execution_wrapper(self, task_data_d, ui_message_fn, ui_progress_fn, ui_task_ended_fn,
                                LogLevels, TaskStatuses):
        try:
            self.status = self.task_execution_fn(self.task_name, task_data_d, ui_message_fn, ui_progress_fn,
                                                 ui_task_ended_fn, LogLevels, TaskStatuses)
        except:
            self.status = TaskStatuses.ERRORED

            stack_trace = traceback.format_exc()
            msg = 'Exception occurred in Task execution function - stack trace follows.\n\n%s' % stack_trace
            self.ui_message_fn(msg, LogLevels.ERROR)

    # parm signature for the ui functions are:
    # 
    #   ui_message_fn(message, message_level)
    #
    #   ui_update_progress_fn(percent_complete)
    #
    #   ui_task_ended_fn(completion_status, completion_message)
    #
    def ui_message_fn(self, message, message_level):

        self._ui_message_driver_fn(self.task_name, message, message_level)

    def ui_update_progress_fn(self, percent_complete):

        self.ui_update_progress_driver_fn(self.task_name, percent_complete)

    def ui_task_ended_fn(self, completion_status, completion_message):

        self.ui_task_ended_driver_fn(self.task_name, completion_status, completion_message)


class Plugin(PluginBase):

    def __init__(self):

        super(Plugin, self).__init__()

        self.task_by_name = {}

    def setup_task(self, task_name, validation_to_start_fn, task_execution_fn):

        # param signatures for task functions are:
        #
        #   validation_to_start_fn(task_name, task_data_d, ui_message_fn, LogLevels)
        #          |
        #          + -> returns bool, True if OK to start Task
        #
        #   task_execution_fn(task_name, task_data_d, ui_message_fn, ui_progress_fn, ui_task_ended_fn,
        #                     LogLevels, TaskStatuses)
        #
        task = Task(task_name, validation_to_start_fn, task_execution_fn,
                    self._ui_message_driver_fn, self._ui_update_progress_driver_fn,
                    self._ui_task_ended_driver_fn)

        self.task_by_name[task_name] = task

    def get_status(self, task_name):

        if task_name not in self.task_by_name:
            self.ui_message_fn('Task name "%s" not registered with TaskManager' % task_name, LogLevels.ERROR)
            return None

        task = self.task_by_name[task_name]
        return task.get_status()

    @register_plugin_op
    def get_registered_tasks(self, op_data):

        registered_task_list = sorted(self.task_by_name.keys())
        self.plugin_to_webbrowser('received_registered_tasks',
                                  {'registered_task_list': registered_task_list})

    @register_plugin_op
    def start_task(self, task_data_d):

        task_name = task_data_d.get('task_name', '')

        if task_name not in self.task_by_name:
            self._ui_message_driver_fn(task_name, 'Task name "%s" not registered with TaskManager' % task_name,
                                       LogLevels.ERROR)
            return

        task = self.task_by_name[task_name]

        if task.status in (TaskStatuses.WORKING, TaskStatuses.PAUSED):
            msg = 'Task "%s" is already running or is paused.' % task_name
            self._ui_message_driver_fn(task_name, msg, LogLevels.WARNING)
            return

        try:
            is_valid_to_start = task.validation_to_start_fn(task_name, task_data_d, task.ui_message_fn, LogLevels)
        except:
            stack_trace = traceback.format_exc()
            msg = 'Exception occurred in Task "%s" validation function - stack trace follows.\n\n%s' % \
                    (task_name, stack_trace)
            self._ui_message_driver_fn(task_name, msg, LogLevels.ERROR)
            return

        if not is_valid_to_start:
            msg = 'Task validation for start has FAILED.'
            self._ui_message_driver_fn(task_name, msg, LogLevels.ERROR)
            return

        self.plugin_to_webbrowser('ui_disable_controls', {'task_name': task_name})

        try:
            self.execution_thread = threading.Thread(target=task._task_execution_wrapper,
                                                     args=(task_data_d,
                                                           task.ui_message_fn,
                                                           task.ui_update_progress_fn,
                                                           task.ui_task_ended_fn,
                                                           LogLevels,
                                                           TaskStatuses))
            self.status = TaskStatuses.WORKING
            self.execution_thread.start()
        except:
            stack_trace = traceback.format_exc()
            msg = 'Exception occurred starting Task thread - stack trace follows.\n\n%s' % stack_trace
            self._ui_message_driver_fn(task_name, msg, LogLevels.ERROR)

            self.plugin_to_webbrowser('ui_enable_controls', {'task_name': task_name})
            return

    def _ui_message_driver_fn(self, task_name, message, message_level):

        self.plugin_to_webbrowser('ui_message',
                                  {'task_name': task_name, 'message': message, 'message_level': message_level})

    def _ui_update_progress_driver_fn(self, task_name, percent_complete):

        self.plugin_to_webbrowser('ui_update_progress',
                                  {'task_name': task_name, 'percent_complete': percent_complete})

    def _ui_task_ended_driver_fn(self, task_name, completion_status, completion_message):

        self.plugin_to_webbrowser('ui_task_ended',
                                  {'task_name': task_name, 'completion_status': completion_status,
                                   'completion_message': completion_message})


